CDCgov/prime-data-hub
# #1069 Added HL7 test framework and decode phone number from HL7

This PR adds:
1. A test framework to compare HL7 files to the expected internal data output generated by the serializer.  This framework allows us to add HL7 files to the test folder with an internal.csv file with the expected values.  The framework runs during the unit tests and generates a unit test on a per HL7 file basis and output the results.  A unit test failure for the given file means there was one or more expected values not met.  There can also be text output on the unit test that includes any schema errors and/or warnings as well as data comparison errors and/or warnings.  A data comparison error occurs when an expected value was added to the internal.csv that is not blank and the HL7 serializer returned a different value including empty.  A data warning occurs when there is no expected value (blank) and the HL7 serializer return a value (e.g. an addition to the schema or newly decoded value). 
2. Decoding the XTN components to extract phone numbers.  I also pass the phone number through the existing Element class phone number check.
3. Testing of the CareEvolution HL7 file.  I went segment by segment then through each schema item.

How I created the internal.csv?  I just printed out the current output of the serializer and pasted it in a file.  I need to create a CLI command to make that easier.

Note the Flagler HL7 file is v2.3

To test:
1. Run `./gradlew test` and verify the ObservationMessageTests ran successfully.
2. To view the results you can go to build/reports/tests/test/classes/gov.cdc.prime.router.serializers.datatests.ObservationMessageTests.html
3. To play with it you can change one or more values in one of the internal.csv files and rerun the tests.
4. To run just that test you could run `./gradlew test --tests gov.cdc.prime.router.serializers.datatests.ObservationMessageTests`

## Changes
- Added test framework
- Decode XTN phone numbers
- Added unit test for phone decoding

## Checklist

### Testing

### Security

### Process

## Security
*List potential security threats and mitigations in the PR*

## Fixes
- #1069

## To Be Done
- Need to add a CLIL command to generate the internal.csv from HL7 to make easier the generation of the expected values.
- Need to support batch HL7 messages when checking for ORU-R01 in the unit test
- Need to investigate what to do with phone extensions
- May need to check if there is a country code available before using the mapper - this implies some smarts to figure out when to use which
- What happens when there are multiple results with multiple OBX segment (e.g. Covid, Influenza A, Influenza B)
